File: src/consoleReader.ts
import * as readline from "readline";

export class ConsoleReader {
  private userInput = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  public async readUserInput(
    query: string,
    callback: (response: string) => void
  ) {
    const enteredText = await new Promise((resolve) =>
      this.userInput.question(query, (response: string) => {
        callback(response);
        resolve(response);
      })
    );

    return enteredText;
  }

  public async prompt() {
    this.userInput.prompt();
  }

  public async onLine(fn: (text: string) => void) {
    this.userInput.on("line", (text: string) => {
      fn(text);
      this.userInput.prompt(true);
    });
  }
}

File: CSharp/ChatApp/ConsoleReader.cs
using System;

namespace ChatApp
{
    public class ConsoleReader
    {
        private readonly System.IO.TextReader input = Console.In;
        private readonly System.IO.TextWriter output = Console.Out;

        public async void ReadUserInput(string query, Action<string> callback)
        {
            output.Write(query);
            var response = await input.ReadLineAsync();
            callback(response);
        }

        public void Prompt()
        {
            output.Flush();
        }

        public void OnLine(Action<string> fn)
        {
            while (true)
            {
                var text = input.ReadLine();
                if (text == null) break;
                fn(text);
            }
        }
    }
}
